### DISCUSSION TITLE
Inline code in figure captions

### DISCUSSION THREAD

**Discussion by Dpananos** [quarto-dev/quarto-cli#1334] [Q&A] [answered]
I'd like to run some R code in a figure caption.  Something like

```{r}
nsims <- 1000
x <- rnorm(nsims)
```

```{r}
#| fig-cap: `r nsims` draws from a standard normal
hist(x)
```

This is doable in R markdown, but quarto seems to error out due to the inline execution.  Can I execute code in figure captions?  If so, what is the right way to do this?

**Comment by mcanouil** [comment]
Hi,

I don't think you can with the chunk option defined in body, since fig-cap must be a string.
(To note, it is not working with knitr/rmarkdown, thus in Quarto.)
But, you can use the old way:

````
---
self-contained: true
format: html
---

```{r}
nsims <- 1000
x <- rnorm(nsims)
```

The following works with knitr/rmarkdown/Quarto:

```{r, fig.cap = paste(nsims, "draws from a standard normal")}
hist(x)
```

The following does not work with knitr/rmarkdown:

```{r}
#| fig-cap: paste(nsims, "draws from a standard normal")
hist(x)
```
````

``#| fig-cap: `r nsims` draws from a standard normal`` is not a valid syntax even for rmarkdown.
```
Error in yaml::yaml.load(meta, handlers = list(expr = parse_only)) : 
  Scanner error: while scanning for the next token at line 1, column 10 found character that cannot start any token at line 1, column 10
```

**Comment by cderv** [comment, marked as answer]
As documented, when using R, you need to use the `!expr` syntax in YAML so that an evaluation happens before YAML parsing 
https://quarto.org/docs/computations/r.html#chunk-options

````markdown
---
self-contained: true
format: html
---

```{r}
nsims <- 1000
x <- rnorm(nsims)
```

The following does not work with knitr/rmarkdown:

```{r}
#| fig-cap: !expr paste(nsims, "draws from a standard normal")
hist(x)
```

````

![image](https://user-images.githubusercontent.com/6791940/177958111-cee21e84-da01-4feb-9c97-8904a6005a91.png)

**Reply by mcanouil** [reply]
`!expr` also works for knitr/rmarkdown (as expected) but I did not see any mention of this anywhere or almost.
The only time I saw this was in @yihui blog post announcing the changes to knitr: [News from knitr v1.35 to v1.37: Alternative Syntax for Chunk Options, and Some New Engines](https://yihui.org/en/2022/01/knitr-news/).

**Reply by cderv** [reply]
> I did not see any mention of this anywhere or almost.

It is only useful with YAML that needs to be evaluated before parsing.  `!expr` is a feature from **yaml** R package, and works also in document header. You'll find mention in the R package doc, in the R Markdown definitive guide (https://bookdown.org/yihui/rmarkdown/output-formats.html, https://bookdown.org/yihui/rmarkdown/html-document.html#custom-fun-dfprint) and probably other post on the web (like with **bslib** support for R Markdown (https://rstudio.github.io/bslib/articles/bslib.html#low-level) 

With knitr you can also use R expression and not YAML when you need R expression as shown in the blog post.

**Reply by mcanouil** [reply]
All the links are only using this feature, but not really describing (focussing on) it unfortunately.
This makes it harder to spot/remember, at least in knitr/rmarkdown documentations, although Quarto documentation highlights this pretty well in fact 😉, I still missed this part 😅.
